### PR TITLE
Allow users to use ruleset in Standalone mode

### PR DESCRIPTION
**Pull request by Evangelink**

### CODE REVIEW

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+
+        private readonly Dictionary<string, ProjectData> projectPathToCachedData = new Dictionary<string, ProjectData>();
+        private readonly Dictionary<string, FileChangeTracker> filesToTrack = new Dictionary<string, FileChangeTracker>();
+
+        private bool isDisposed;
+
+        [ImportingConstructor]
+        public ProjectsRuleSetProvider(IHost host)
+            : this(host.GetService<IProjectSystemHelper>(), host.GetService<SVsFileChangeEx, IVsFileChangeEx>(),
+                  host.GetMefService<IActiveSolutionTracker>(), InMemoryConfigurationProvider.Instance, host.Logger)
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/ProjectsRuleSetProvider.cs]: At present the _InMemoryConfigurationProvider_ will only accurately track if we are in connected more or not, but I guess that's ok for this class.

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+        {
+            filesToTrack.Values.ToList().ForEach(x => x.UpdatedOnDisk -= OnUpdatedOnDisk);
+            filesToTrack.Clear();
+
+            projectPathToCachedData.Clear();
+        }
+
+        private async Task BuildCacheAsync()
+        {
+            while (!projectSystemHelper.IsSolutionFullyOpened())
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/ProjectsRuleSetProvider.cs]: Is this _while_ needed? I don't think the active solution tracker will raise the event until the solution has finished loading, so it shouldn't be needed when responding to those events.
That leaves the initial construction; if there isn't a solution fully-open at the time this class is created, then couldn't we rely on the event from the active solution tracker in that case too?

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+            {
+                await Task.Delay(2000); // REVIEW: Is 2s a good enough time?
+            }
+
+            projectSystemHelper.GetSolutionProjects()
+                .ToList()
+                .ForEach(p => projectPathToCachedData.Add(p.FullName, ProcessProjectAndCollectFilesToTrack(p)));
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/ProjectsRuleSetProvider.cs]: We should really add an extension method that gives us a _ForEach_ over an enumeration so we don't always have to put it in a list first.

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+        }
+
+        /// <remarks>
+        /// [SIDE EFFECT] The set of ruleset is update at each call.
+        /// </summaryremarks>
+        private ProjectData ProcessProjectAndCollectFilesToTrack(EnvDTE.Project project)
+        {
+            TrackIfNotTracked(project.FullName);
+
+            var projectRuleSetPath = projectSystemHelper.GetProjectProperty(project, Constants.CodeAnalysisRuleSetPropertyKey);
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/ProjectsRuleSetProvider.cs]: ?doesn't handle ruleset path being null/empty?

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+            var projectRuleSetFullPath = GetFullPath(projectRuleSetPath, projectDirectoryFullPath);
+
+            var hasAnySonarRule = false;
+
+            if (File.Exists(projectRuleSetFullPath))
+            {
+                var projectRuleSet = RuleSet.LoadFromFile(projectRuleSetFullPath);
+
+                // 1. Collect all paths (current ruleset + includes)
+                var ruleSetIncludeFullPaths = projectRuleSet.RuleSetIncludes
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/ProjectsRuleSetProvider.cs]: An included ruleset can include other rulesets, can't it?

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+            {
+                await Task.Delay(2000); // REVIEW: Is 2s a good enough time?
+            }
+
+            projectSystemHelper.GetSolutionProjects()
+                .ToList()
+                .ForEach(p => projectPathToCachedData.Add(p.FullName, ProcessProjectAndCollectFilesToTrack(p)));
+
+            // Subscribe to file changes for the projects and the rulesets
+            filesToTrack.Values.ToList().ForEach(x => x.UpdatedOnDisk += OnUpdatedOnDisk);
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/ProjectsRuleSetProvider.cs]: Hasn't _TrackIfNotTracked_ already been called for all of these files?

Ah; _TrackIfNotTracked_ creates the tracking wrapper and adds it to the collection, but this wires up the event from the tracking wrapper. Could both steps be done in _TrackIfNotTracked_?

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+        private async void OnUpdatedOnDisk(object sender, EventArgs e)
+        {
+            var fileChangeTracker = sender as FileChangeTracker;
+            if (fileChangeTracker == null)
+            {
+                return;
+            }
+
+            if (!fileChangeTracker.FilePath.EndsWith(".ruleset"))
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/ProjectsRuleSetProvider.cs]: I'm not sure if a ruleset _has_ to have a _.ruleset_ extension; I suspect VS and MSBuild would handle it correctly if it didn't, but I've never seen a ruleset file with a different extension. Not worth worrying about.

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+            {
+                var cachedProjectRuleSetPath = projectPathToCachedData[fileChangeTracker.FilePath].RuleSetPath;
+
+                var project = projectSystemHelper.GetSolutionProjects().FirstOrDefault(p =>
+                    p.FullName.Equals(fileChangeTracker.FilePath, StringComparison.OrdinalIgnoreCase));
+                if (project != null)
+                {
+                    var projectRuleSetPath = projectSystemHelper.GetProjectProperty(project, Constants.CodeAnalysisRuleSetPropertyKey);
+                    var projectDirectoryFullPath = new FileInfo(project.FullName).Directory.FullName;
+                    var projectRuleSetFullPath = GetFullPath(projectRuleSetPath, projectDirectoryFullPath);
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/ProjectsRuleSetProvider.cs]: 181-183 - candidate for a function? I think this logic occurs more than once.

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+            {
+                add
+                {
+                    lock (lockThis)
+                    {
+                        updatedOnDisk += value;
+                    }
+                    subscriptionCount++;
+
+                    if (subscriptionCount == 1)
```
Comment (duncanp-sonar) [inline src/Integration/MefServices/ProjectsRuleSetProvider.cs]: _Debug.Assert_ that is 1

**Comment by duncanp-sonar** [comment]
Couldn't think of any show-stopping issues, and I don't think there's anything major in the comments I added.

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs (hunk 2 above)
Comment (Evangelink) [reply src/Integration/MefServices/ProjectsRuleSetProvider.cs]: On all my tests the solution was never fully loaded before the ctor was called. The active solution tracker event was also not always raising at the expected time hence this wait.

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs (hunk 6 above)
Comment (Evangelink) [reply src/Integration/MefServices/ProjectsRuleSetProvider.cs]: That's a good suggestion. The code was really different at first and I haven't seen this simplification when trying to improve it.

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs (hunk 7 above)
Comment (Evangelink) [reply src/Integration/MefServices/ProjectsRuleSetProvider.cs]: I actually wanted to start with the other way (i.e. detecting we are called-back for a project but I don't have the info).

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs (hunk 5 above)
Comment (Evangelink) [reply src/Integration/MefServices/ProjectsRuleSetProvider.cs]: Good point. That might be tricky to handle though...

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs (hunk 1 above)
Comment (Evangelink) [reply src/Integration/MefServices/ProjectsRuleSetProvider.cs]: Yes we only care about being in the new connected mode.

File: src/Integration/MefServices/ProjectsRuleSetProvider.cs (hunk 8 above)
Comment (Evangelink) [reply src/Integration/MefServices/ProjectsRuleSetProvider.cs]: It's indeed called twice but creating a method returning 3 parameters is not ideal. Creating a class with 3 properties readonly for that is overkill in c#.

**Pull request merged**